subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt

```diff
@@ -0,0 +1,176 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.normalization
+
+import kotlinx.metadata.Flag
+import kotlinx.metadata.KmDeclarationContainer
+import kotlinx.metadata.jvm.KotlinClassHeader
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import kotlinx.metadata.jvm.signature
+import org.gradle.internal.normalization.java.ApiClassExtractor
+import org.gradle.internal.normalization.java.impl.AnnotationMember
+import org.gradle.internal.normalization.java.impl.ApiMemberWriter
+import org.gradle.internal.normalization.java.impl.ArrayAnnotationValue
+import org.gradle.internal.normalization.java.impl.ClassMember
+import org.gradle.internal.normalization.java.impl.FieldMember
+import org.gradle.internal.normalization.java.impl.InnerClassMember
+import org.gradle.internal.normalization.java.impl.MethodMember
+import org.gradle.internal.normalization.java.impl.MethodStubbingApiMemberAdapter
+import org.gradle.internal.normalization.java.impl.SimpleAnnotationValue
+import org.objectweb.asm.ClassReader
+import org.objectweb.asm.ClassVisitor
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.MethodVisitor
+import org.objectweb.asm.Opcodes
+
+
+class KotlinApiClassExtractor : ApiClassExtractor(
+    emptySet(),
+    { classReader, classWriter ->
+        KotlinApiMemberWriter(
+            MethodStubbingApiMemberAdapter(classWriter),
+            MethodCopyingApiMemberAdapter(classReader, classWriter)
+        )
+    }
+)
+
+
+private
+class KotlinApiMemberWriter(apiMemberAdapter: ClassVisitor, val inlineMethodWriter: MethodCopyingApiMemberAdapter) : ApiMemberWriter(apiMemberAdapter) {
+
+    val kotlinMetadataAnnotationSignature = "Lkotlin/Metadata;"
+
+    val inlineFunctions: MutableSet<String> = HashSet()
+    val internalFunctions: MutableSet<String> = HashSet()
+
+    override fun writeClass(classMember: ClassMember, methods: Set<MethodMember>, fields: Set<FieldMember>, innerClasses: Set<InnerClassMember>) {
+        classMember.annotations.firstOrNull {
+            it.name == kotlinMetadataAnnotationSignature
+        }?.let {
+            when (val kotlinMetadata = KotlinClassMetadata.read(parseKotlinClassHeader(it))) {
+                is KotlinClassMetadata.Class -> kotlinMetadata.toKmClass().extractFunctionMetadata()
+                is KotlinClassMetadata.FileFacade -> kotlinMetadata.toKmPackage().extractFunctionMetadata()
+                else -> {
+                    // KotlinClassMetadata.SyntheticClass || KotlinClassMetadata.MultiFileClassFacade || KotlinClassMetadata.MultiFileClassPart || KotlinClassMetadata.Unknown
```

Comment:
❓ I wonder if we should try to support more, fallback to compiling, or fail on some unsupported metadata types.

`MultiFileClassFacade` and `MultiFileClassPart` could happen if `@JvmMultifileClass` is used in build logic code.

`SyntheticClass` could happen if `@JvmSynthetic` is used in build logic code.

`Unknown` could happen if, for example, dependencies were compiled with a _future_ Kotlin version that introduce new metadata.

This could be addressed in a subsequent PR.

Reply:
Here's a fix for `MultiFileClassFacade` and `MultiFileClassPart` - missed one case there - https://github.com/gradle/gradle/pull/14827

Reply:
And for `Unknown` metadata - fallback to using the full hash of the artifact: https://github.com/gradle/gradle/pull/14834